### PR TITLE
update: change the default version of OpenFunction to the latest stable version

### LINKED ISSUE #20: Need to add some useful options to improve the efficiency of cli
`--without`, for quick reverse screening of certain components
`--force`, for skipping the confirm phase

### LINKED ISSUE #21: Replace the default OpenFunction version installed with latest stable release
Currently, `ofn install --all` will install v0.4.0 as the default version which should be changed to [the latest stable release](https://api.github.com/repos/OpenFunction/OpenFunction/releases/latest).

Another issue is default installation timeout could be adjusted to 15m

### PR DESCRIPTION
**Pull request by tpiperatgod**
1. change the default version of OpenFunction to the latest stable version
2. add the `--force` option to force the operation
3. use the spinner instead of the original process display
4. adjust the function hierarchy to make some functions more generic
5. update docs

### CODE REVIEW

**Comment by tpiperatgod** [comment]
Untitled

File: README.md [hunk 1]
```diff
@@ -29,51 +29,56 @@ Run `ofn install --all` to implement a simple deployment. By default, this comma
...
-🚀 Completed in 2m3.638035129s.
+ ✓ Dapr - Completed!
+ ✓ Keda - Completed!
+ ✓ Knative Serving - Completed!
+ ✓ Shipwright - Completed!
+ ✓ Cert Manager - Completed!
+ ✓ Ingress - Completed!
+ ✓ OpenFunction - Completed!
+🚀 Completed in 4m45.512145978s.
+Notice that you are using Knative runtime, you can refer to the following to configure Knative's network layer (Assuming you are using Kourier) and DNS.
```
Comment (benjaminhuo) [inline README.md]: you are using Knative runtime => If you are using Knative runtime

you can refer to the following => you can refer to the following instructions

File: README.md [hunk 2]
```diff
@@ -29,51 +29,56 @@ Run `ofn install --all` to implement a simple deployment. By default, this comma
...
+Where 1.2.3.4 can be replaced by your node address or loadbalancer address:
```
Comment (benjaminhuo) [inline README.md]: can be => should be

File: README.md [hunk 3]
```diff
@@ -29,51 +29,56 @@ Run `ofn install --all` to implement a simple deployment. By default, this comma
...
+
```
Comment (benjaminhuo) [inline README.md]: Refer to the `Fetch the External IP address or CNAME by running the command:` section to adjust the `1.2.3.4` desc.

As I can remember, the loadbalancer's IP is assigned automatically, why patching this IP?

Patching a node's IP to this only works for single node scenario?

https://knative.dev/docs/install/serving/install-serving-with-yaml/
image

File: docs/install.md [hunk 4]
```diff
@@ -17,6 +17,7 @@ This command will help you to install OpenFunction and its dependencies.
 --shipwright         For installing ShipWright.
 --sync               For installing OpenFunction Sync Runtime (To be supported).
 --upgrade            Upgrade components to target version while installing.
+--force              Forced installation.
```
Comment (benjaminhuo) [inline docs/install.md]: `Forced installation.` is not clear enough.
Actually, this param can be used during the install and uninstall process and it's used to ignore entering y to confirm?

So it's similar to `apt-get -y install` :
Automatic yes to prompts; assume "yes" as answer to all prompts and run non-interactively.
https://askubuntu.com/questions/672892/what-does-y-mean-in-apt-get-y-install-command

File: docs/uninstall.md [hunk 5]
```diff
@@ -54,11 +55,10 @@ ofn uninstall --all --wait
 
 ### Uninstall a specific version of OpenFunction
 
-> Default to v0.4.0 if no version specified
+> Default to the version of the OpenFunction exists
```
Comment (benjaminhuo) [inline docs/uninstall.md]: Default to the version of the OpenFunction exists 
=>
Default to the version of the OpenFunction currently installed

File: pkg/cmd/subcommand/install.go [hunk 6]
```diff
@@ -13,24 +15,24 @@ import (
...
-	"golang.org/x/sync/errgroup"
+	"k8s.io/apimachinery/pkg/util/version"
 	"k8s.io/cli-runtime/pkg/genericclioptions"
 	k8s "k8s.io/client-go/kubernetes"
 )
 
-var (
-	w io.Writer
+const (
+	openFunctionLatestReleaseUrl = "https://api.github.com/repos/OpenFunction/OpenFunction/releases/latest"
```
Comment (benjaminhuo) [inline pkg/cmd/subcommand/install.go:32]: can be optimized for Chinese users with https://www.jsdelivr.com/?
Looks like this API can not be converted and it's not blocked by default?

File: pkg/cmd/subcommand/install.go [hunk 7]
```diff
@@ -161,17 +167,16 @@ func (i *Install) ValidateArgs(cmd *cobra.Command, args []string) error {
...
 	operator := common.NewOperator(runtime.GOOS, i.OpenFunctionVersion, i.Timeout, i.RegionCN, i.Verbose)
-	ti := util.NewTaskInformer("")
 	continueFunc := func() bool {
 		reader := bufio.NewReader(os.Stdin)
-		fmt.Fprintln(w, ti.BeforeTask("You have used the `--upgrade` parameter, which means that the installation process "+
-			"will overwrite the components that already exist.\n"+
-			"Please ensure that you understand the meaning of this command "+
-			"and follow the prompts below to confirm the action.\n"+
-			"Enter 'y' to continue and 'n' to abort:"))
+		util.BeforeTask("You have used the `--upgrade` parameter, which means that the installation process " +
```
Comment (benjaminhuo) [inline pkg/cmd/subcommand/install.go]: "You have specified the `--upgrade` flag, which means that the installation process " +
			"will upgrade components currently installed.\n" +
			"Please make sure that you're aware of the consequences of this command " +
			"and follow the prompts below to confirm the upgrade.\n" +
			"Enter 'y' to continue and 'n' to abort:"

File: pkg/cmd/subcommand/install.go [hunk 8]
```diff
@@ -211,14 +216,18 @@ func (i *Install) RunInstall(cl *k8s.Clientset, cmd *cobra.Command) error {
 	operator.Inventory = inventoryPending
 	inventoryExist := getExistComponentsInventory(ctx, cl)
 
-	fmt.Fprintln(w, ti.BeforeTask("Start installing OpenFunction and its dependencies.\n"+
-		"Here are the components and corresponding versions to be installed:"))
+	util.BeforeTask("Start installing OpenFunction and its dependencies.\n" +
+		"Here are the components and corresponding versions to be installed:")
```
Comment (benjaminhuo) [inline pkg/cmd/subcommand/install.go]: Here are the components and corresponding versions to be installed:
=>
The following components will be installed:

File: pkg/cmd/subcommand/install.go [hunk 9]
```diff
@@ -211,14 +216,18 @@ func (i *Install) RunInstall(cl *k8s.Clientset, cmd *cobra.Command) error {
...
 	printInventory(inventory.GetVersionMap(inventoryPending))
-	if !reflect.DeepEqual(inventoryExist, map[string]bool{}) {
-		fmt.Fprintln(w, ti.BeforeTask("The following components already exist:"))
+	if !reflect.DeepEqual(inventoryExist, map[string]bool{}) && !i.Force {
+		if !i.Upgrade {
+			util.BeforeTask("The following components already exist (will be skipped):")
```
Comment (benjaminhuo) [inline pkg/cmd/subcommand/install.go]: The following components already exist (will be skipped):
=>
The following existing components will be skipped:

File: pkg/cmd/subcommand/install.go [hunk 10]
```diff
@@ -211,14 +216,18 @@ func (i *Install) RunInstall(cl *k8s.Clientset, cmd *cobra.Command) error {
...
+		} else {
+			util.BeforeTask("The following components already exist:")
```
Comment (benjaminhuo) [inline pkg/cmd/subcommand/install.go]: The following components already exist:
=>
The following existing components will be upgraded:

File: pkg/cmd/util/spinner/spinner.go [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+	return msg
+}
+
+// UpdateMessage updates the spinner message
+func (s *Spinner) UpdateMessage(message string) {
+	s.message.SetValue(s.handleMessage(message))
+}
+
+// Success marks spinner as success
+func (s *Spinner) Success() {
```
Comment (benjaminhuo) [inline pkg/cmd/util/spinner/spinner.go]: Success => Done

File: pkg/cmd/util/spinner/spinner.go [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+	s.UpdateMessage(completed)
+	s.stop(successStatus)
+}
+
+// SuccessWithMessage marks spinner as success and update message
+func (s *Spinner) SuccessWithMessage(message string) {
```
Comment (benjaminhuo) [inline pkg/cmd/util/spinner/spinner.go]: Done(message)

File: pkg/cmd/util/spinner/spinner.go [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+	s.UpdateMessage(failed)
+	s.stop(errorStatus)
+	if err != nil {
+		s.group.err = err
+		s.group.errC <- err
+	}
+}
+
+// ErrorWithMessage marks spinner as error and update message
+func (s *Spinner) ErrorWithMessage(message string, err error) {
```
Comment (benjaminhuo) [inline pkg/cmd/util/spinner/spinner.go]: Can still use Error with different args

File: pkg/cmd/util/spinner/spinner.go [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+	}
+}
+
+func (s *Spinner) stop(status int) {
+	s.status.SetValue(status)
+	s.group.redraw()
+	s.group.Done()
+}
+
+func (s *Spinner) sprint() string {
```
Comment (benjaminhuo) [inline pkg/cmd/util/spinner/spinner.go]: sprint => refresh?

File: pkg/cmd/util/spinner/spinner_group.go [hunk 15]
```diff
@@ -0,0 +1,167 @@
+package multispinner
```
Comment (benjaminhuo) [inline pkg/cmd/util/spinner/spinner_group.go]: Can we just call multispinner spinnerlib?

File: pkg/cmd/util/spinner/spinner_group.go [hunk 16]
```diff
@@ -0,0 +1,167 @@
...
+			case <-ctx.Done():
+				g.stopSpinnersWithError(ctx.Err().Error())
+			case <-g.errC:
+				g.stopSpinnersWithError("stopped")
+			}
+		}
+	}()
+}
+
+func (g *SpinnerGroup) stopSpinnersWithError(errMsg string) {
```
Comment (benjaminhuo) [inline pkg/cmd/util/spinner/spinner_group.go]: Can just be stop(errMsg)

**Review by benjaminhuo — CHANGES_REQUESTED** [review]

File: pkg/cmd/util/spinner/spinner.go [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+	if s.name != nil {
+		msg = fmt.Sprintf("%s - %s", *s.name, message)
+	} else {
+		msg = message
+	}
+	return msg
+}
+
+// UpdateMessage updates the spinner message
+func (s *Spinner) UpdateMessage(message string) {
```
Comment (benjaminhuo) [inline pkg/cmd/util/spinner/spinner.go]: UpdateMessage can be change to UpdateProcess or simply Update/Refresh

File: README.md (hunk 1 above)
Comment (tpiperatgod) [reply README.md]: I think these tips can be removed...

File: README.md (hunk 3 above)
Comment (tpiperatgod) [reply README.md]: Yes, as there are many configuration options for this section, I think the tips can be removed...

File: docs/install.md (hunk 4 above)
Comment (tpiperatgod) [reply docs/install.md]: Change `--force` to `--yes` and `-y` as a short form.

File: pkg/cmd/subcommand/install.go (hunk 6 above)
Comment (tpiperatgod) [reply pkg/cmd/subcommand/install.go:32]: yes, I have tested this before and it works

File: pkg/cmd/util/spinner/spinner.go (hunk 13 above)
Comment (tpiperatgod) [reply pkg/cmd/util/spinner/spinner.go]: change to this:
```go
// Error marks spinner as error
func (s *Spinner) Error(err error) {
	s.ErrorWithMessage(failed, err)
}

// ErrorWithMessage marks spinner as error and update message
func (s *Spinner) ErrorWithMessage(message string, err error) {
	s.Update(message)
	s.stop(errorStatus)
	if err != nil {
		s.group.err = err
		s.group.errC <- err
	}
}
```

File: pkg/cmd/util/spinner/spinner_group.go (hunk 15 above)
Comment (tpiperatgod) [reply pkg/cmd/util/spinner/spinner_group.go]: Is it ok to change `multispinner` to `terminalspinner`?

Comment (benjaminhuo) [reply pkg/cmd/util/spinner/spinner_group.go]: or just spinners

File: VERSION [hunk 18]
```diff
@@ -1 +1 @@
-latest
```
Comment (benjaminhuo) [inline VERSION]: The main branch should have latest flag, while the release branch should have v0.5.0

File: pkg/cmd/util/spinner/spinner_group.go (hunk 15 above)
Comment (tpiperatgod) [reply pkg/cmd/util/spinner/spinner_group.go]: If I use `spinner*` as the package name, the IDE has this message: "Name starts with the package name", does it matter?

File: VERSION (hunk 18 above)
Comment (tpiperatgod) [reply VERSION]: Updated to `latest` in the second commit

File: README.md [hunk 19]
```diff
@@ -85,41 +79,32 @@ Run `ofn uninstall --all` to uninstall OpenFunction and its dependencies (or jus
 For more information, refer to the [ofn uninstall document](docs/uninstall.md).
 
 ```shell
-~# ofn uninstall --all
+~# ofn uninstall --all --force
```
Comment (benjaminhuo) [inline README.md]: still force here?

File: pkg/cmd/subcommand/uninstall.go [hunk 20]
```diff
@@ -105,8 +106,9 @@ ofn uninstall --all --version v0.4.0
 	cmd.Flags().BoolVar(&i.RegionCN, "region-cn", false, "For users who have limited access to gcr.io or github.com.")
 	cmd.Flags().BoolVar(&i.DryRun, "dry-run", false, "Used to prompt for the components and their versions to be uninstalled by the current command.")
 	cmd.Flags().BoolVar(&i.WaitForCleared, "wait", false, "Awaiting the results of the uninstallation.")
-	cmd.Flags().StringVar(&i.OpenFunctionVersion, "version", "v0.4.0", "Used to specify the version of OpenFunction to be uninstalled.")
-	cmd.Flags().DurationVar(&i.Timeout, "timeout", 5*time.Minute, "Set timeout time. Default is 5 minutes.")
+	cmd.Flags().BoolVarP(&i.Yes, "force", "y", false, "Automatic yes to prompts.")
```
Comment (benjaminhuo) [inline pkg/cmd/subcommand/uninstall.go]: Still force?

File: pkg/cmd/util/spinner/spinner_group.go (hunk 15 above)
Comment (benjaminhuo) [reply pkg/cmd/util/spinner/spinner_group.go]: What if changing `spinner/*.go  to spinners/*.go` and then using spinners as pkg name?

**Review by benjaminhuo — APPROVED** [review]

**Pull request merged**